### PR TITLE
fix!: make params matching use express logic

### LINKED ISSUE #66: Incorrect path handling when params not provided
- When you have a route with a `/:foo` path and you don't pass the the `foo` param when making a request then a request with a literal `/:foo` path will be made. This seems non-ideal. The param is required so the error should probably be more explicit and triggered before the request is made?
- The module doesn't seem to declare support for optional params (`/:foo?`) but I think it should be able to support those. This kinda ties to the previous issue.

### PR DESCRIPTION
**Pull request by rchl**
Resolves #66

(at least partially)

### CODE REVIEW

**Review by ezypeeze — DISMISSED** [review]

**Comment by ezypeeze** [comment]
All good for me, this way is much better.

File: lib/plugins/api.template.js [hunk 1]
```diff
@@ -61,14 +45,12 @@ function generateAPI(controllerMapping, ctx) {
     for (const key of Object.keys(controllerMapping)) {
         const context = controllerMapping[key];
         if (context && context.path && context.verb) {
+            const path = context.path.replace(/\/$/, '');
```
Comment (rchl) [inline lib/plugins/api.template.js:48]: Not sure what was the reason for this but I would think it's no longer needed when using `path-to-regexp`. Tests still pass.

**Comment by rchl** [comment]
I've adopted changes to tests from your PR and added more tests to test both server and client-side with optional params.

**Comment by rchl** [comment]
Oh, I also wanted the encode function to be customizable through options but I ran out of time for now.

File: lib/plugins/api.template.js (hunk 1 above)
Comment (ezypeeze) [reply lib/plugins/api.template.js]: I think it was to remove a possible '/' on the end of the path

**Review by ezypeeze — APPROVED** [review]

**Comment by rchl** [comment]
Can add the customization in a separate PR.

**Comment by rchl** [comment]
Actually express.js uses `decodeURIComponent` for decoding params so don't need the encoding function to be customizable as there will be parity already as long as express.js is used as a backend.

**Pull request merged**